worker/instancemutater/mutater.go

```diff
@@ -128,20 +133,39 @@ func (m mutaterMachine) watchProfileChangesLoop(profileChangeWatcher watcher.Not
 			if err != nil {
 				return errors.Trace(err)
 			}
-			if err = m.processMachineProfileChanges(info); err != nil {
+			if err = m.processMachineProfileChanges(info); err != nil && errors.IsNotValid(err) {
+				// Return to stop mutating the machine, but no need to restart
+				// the worker.
+				return nil
+			} else if err != nil {
+				return errors.Trace(err)
+			}
+		case <-removed:
+			if err := m.machineApi.Refresh(); err != nil {
 				return errors.Trace(err)
 			}
+			if m.machineApi.Life() == params.Dead {
+				return nil
+			}
 		}
 	}
 }
 
 func (m mutaterMachine) processMachineProfileChanges(info *instancemutater.UnitProfileInfo) error {
-	m.logger.Tracef("%s.processMachineProfileChanges(%#v)", m.id, info)
 	if len(info.CurrentProfiles) == 0 && len(info.ProfileChanges) == 0 {
 		// no changes to be made, return now.
 		return nil
 	}
 
+	if err := m.machineApi.Refresh(); err != nil {
+		return err
+	}
+	if m.machineApi.Life() == params.Dead {
+		// Machine is dead, continue onwards as we can't do anything in this
```

Comment:
I think the comment at the call site is sufficient to indicate what is happening.
This reads a bit like the behaviour changed at some point, but the comment is vestigial. I'd remove it.

Reply:
removed